core/src/main/java/org/apache/accumulo/core/fate/zookeeper/ServiceLock.java

```diff
@@ -762,4 +762,26 @@ public static boolean deleteLock(ZooReaderWriter zk, ServiceLockPath path, Strin
 
     return false;
   }
+
+  /**
+   * Checks that the lock still exists in ZooKeeper. The typical mechanism for determining if a lock
+   * is lost depends on a Watcher set on the lock node. There exists a case where the Watcher may
+   * not get called if another Watcher is stuck waiting on I/O or otherwise hung. In the case where
+   * this method returns false, then the typical action is to exit the server process.
+   *
+   * @return true if lock path still exists, false otherwise and on error
+   */
+  public boolean verifyLockAtSource() {
+    if (getLockPath() == null) {
+      // lock not set yet
+      return false;
+    }
+    try {
+      return null != this.zooKeeper.exists(getLockPath(), false);
```

Comment:
`this.zooKeeper` is almost certainly coming from a shared ZooKeeper instance created from ZooSession and is reused by other parts of the system with other watchers. I'm not sure if that matters here, since this isn't setting a watcher.

This ambiguity over where the ZooKeeper is being created and how it is reused is something I'm addressing in #5124 and am nearly done. I was not expecting to make the changes in 2.1, though. I'm only targeting 3.1. I don't know if it matters for this, but I thought it might make sense to point out, in case you were expecting this to be a unique ZK client.

Reply:
I figured that it did not matter if I was using a shared ZK client in this case as I'm not setting or relying on a watcher and instead checking to see if the node exists on the server.

Reply:
Okay, that's reasonable. I wasn't sure if the watcher thread would have any impact on other operations.